src/validation/rules/NoUnusedVariables.js

```diff
@@ -23,38 +24,53 @@ export function unusedVariableMessage(
 /**
  * No unused variables
  *
- * A GraphQL operation is only valid if all variables defined by an operation
- * are used, either directly or within a spread fragment.
+ * A GraphQL definition is only valid if all variables defined by that
+ * definition are used, either directly or within a spread fragment.
+ *
+ * NOTE: if experimentalFragmentVariables are used, then fragments with
+ * variables defined are considered independent "executable definitions".
+ * So `query Foo` must not define `$a` when `$a` is only used inside
+ * `fragment FragA($a: Type)`
  */
 export function NoUnusedVariables(context: ValidationContext): ASTVisitor {
   let variableDefs = [];
 
-  return {
-    OperationDefinition: {
-      enter() {
-        variableDefs = [];
-      },
-      leave(operation) {
-        const variableNameUsed = Object.create(null);
-        const usages = context.getRecursiveVariableUsages(operation);
-        const opName = operation.name ? operation.name.value : null;
-
-        usages.forEach(({ node }) => {
-          variableNameUsed[node.name.value] = true;
-        });
-
-        variableDefs.forEach(variableDef => {
-          const variableName = variableDef.variable.name.value;
-          if (variableNameUsed[variableName] !== true) {
-            context.reportError(
-              new GraphQLError(unusedVariableMessage(variableName, opName), [
-                variableDef,
-              ]),
-            );
-          }
-        });
-      },
+  const executableDefinitionVisitor = {
+    enter(definition: ExecutableDefinitionNode) {
+      if (!context.isExecutableDefinitionWithVariables(definition)) {
+        return;
+      }
+      variableDefs = [];
     },
+    leave(definition: ExecutableDefinitionNode) {
+      if (!context.isExecutableDefinitionWithVariables(definition)) {
```

Comment:
just check that `variableDefs` is not empty since if it's empty it couldn't be any unused variables.